---
title: "Özel Durumlar ve Özel Durum İşleme (C# Programlama Kılavuzu)"
ms.date: 07/20/2015
ms.prod: .net
ms.technology: devlang-csharp
ms.topic: article
helpviewer_keywords:
- exception handling [C#]
- exceptions [C#]
- C# language, exceptions
ms.assetid: 0001887f-4fa2-47e2-8034-2819477e2344
caps.latest.revision: "33"
author: BillWagner
ms.author: wiwagn
ms.openlocfilehash: 458f6770a89f28dce1e441cd38ec9a56a1c58bc1
ms.sourcegitcommit: 4f3fef493080a43e70e951223894768d36ce430a
ms.translationtype: MT
ms.contentlocale: tr-TR
ms.lasthandoff: 11/21/2017
---
# <a name="exceptions-and-exception-handling-c-programming-guide"></a><span data-ttu-id="8bfea-102">Özel Durumlar ve Özel Durum İşleme (C# Programlama Kılavuzu)</span><span class="sxs-lookup"><span data-stu-id="8bfea-102">Exceptions and Exception Handling (C# Programming Guide)</span></span>
<span data-ttu-id="8bfea-103">C# dil özel durum özellikleri Yardım işleme ile bir program çalışırken oluşan beklenmeyen veya olağanüstü durumlara ilgilenir.</span><span class="sxs-lookup"><span data-stu-id="8bfea-103">The C# language's exception handling features help you deal with any unexpected or exceptional situations that occur when a program is running.</span></span> <span data-ttu-id="8bfea-104">Özel durum işleme kullanan `try`, `catch`, ve `finally` başarılı olmayabilir Eylemler denemek için bunu yapmak için ve daha sonra kaynakları temizlemek için makul karar verirken hataları işlemek için anahtar sözcükler.</span><span class="sxs-lookup"><span data-stu-id="8bfea-104">Exception handling uses the `try`, `catch`, and `finally` keywords to try actions that may not succeed, to handle failures when you decide that it is reasonable to do so, and to clean up resources afterward.</span></span> <span data-ttu-id="8bfea-105">Özel durumlar ortak dil çalışma zamanı tarafından (CLR), .NET Framework veya herhangi bir üçüncü taraf kitaplıklar veya uygulama kodu tarafından oluşturulabilir.</span><span class="sxs-lookup"><span data-stu-id="8bfea-105">Exceptions can be generated by the common language runtime (CLR), by the .NET Framework or any third-party libraries, or by application code.</span></span> <span data-ttu-id="8bfea-106">Özel durumlar kullanarak oluşturulur `throw` anahtar sözcüğü.</span><span class="sxs-lookup"><span data-stu-id="8bfea-106">Exceptions are created by using the `throw` keyword.</span></span>  
  
 <span data-ttu-id="8bfea-107">Çoğu durumda, bir özel durum kodunuzu doğrudan çağrılan olmayan bir yöntemi tarafından oluşturulan, ancak başka bir yöntem kullanarak aşağı çağrı yığınında daha fazla.</span><span class="sxs-lookup"><span data-stu-id="8bfea-107">In many cases, an exception may be thrown not by a method that your code has called directly, but by another method further down in the call stack.</span></span> <span data-ttu-id="8bfea-108">Bu gerçekleştiğinde, CLR yığını bir yöntemle arayan bırakma bir `catch` belirli özel durum türünü ve bu ilk yürütecek için engelleme gibi `catch` olması durumunda engelleme bulur.</span><span class="sxs-lookup"><span data-stu-id="8bfea-108">When this happens, the CLR will unwind the stack, looking for a method with a `catch` block for the specific exception type, and it will execute the first such `catch` block that if finds.</span></span> <span data-ttu-id="8bfea-109">Hayır uygun bulursa `catch` herhangi bir yere çağrı yığınında engellemek, işlemi sonlandırmamız ve kullanıcıya bir ileti görüntüler.</span><span class="sxs-lookup"><span data-stu-id="8bfea-109">If it finds no appropriate `catch` block anywhere in the call stack, it will terminate the process and display a message to the user.</span></span>  
  
 <span data-ttu-id="8bfea-110">Bu örnekte, bir yöntem sıfıra bölme için test ve hata yakalar.</span><span class="sxs-lookup"><span data-stu-id="8bfea-110">In this example, a method tests for division by zero and catches the error.</span></span> <span data-ttu-id="8bfea-111">Özel durum işleme olmadan bu programı ile sonlandırmak bir **DivideByZeroException işlenmemiş** hata.</span><span class="sxs-lookup"><span data-stu-id="8bfea-111">Without the exception handling, this program would terminate with a **DivideByZeroException was unhandled** error.</span></span>  
  
 [!code-csharp[csProgGuideExceptions#18](../../../csharp/programming-guide/exceptions/codesnippet/CSharp/exceptions-and-exception-handling_1.cs)]  
  
## <a name="exceptions-overview"></a><span data-ttu-id="8bfea-112">Özel durumlar genel bakış</span><span class="sxs-lookup"><span data-stu-id="8bfea-112">Exceptions Overview</span></span>  
 <span data-ttu-id="8bfea-113">Özel durumlar aşağıdaki özelliklere sahiptir:</span><span class="sxs-lookup"><span data-stu-id="8bfea-113">Exceptions have the following properties:</span></span>  
  
-   <span data-ttu-id="8bfea-114">Özel durumları olan tüm sonuçta öğesinden türetilen türler `System.Exception`.</span><span class="sxs-lookup"><span data-stu-id="8bfea-114">Exceptions are types that all ultimately derive from `System.Exception`.</span></span>  
  
-   <span data-ttu-id="8bfea-115">Kullanım bir `try` özel durumlar oluşturma deyimleri geçici bloğu.</span><span class="sxs-lookup"><span data-stu-id="8bfea-115">Use a `try` block around the statements that might throw exceptions.</span></span>  
  
-   <span data-ttu-id="8bfea-116">Bir özel durum oluşması sonra `try` engellemek, akışını herhangi bir yere çağrı yığınında mevcut ilk ilişkili özel durum işleyici denetimi atlar.</span><span class="sxs-lookup"><span data-stu-id="8bfea-116">Once an exception occurs in the `try` block, the flow of control jumps to the first associated exception handler that is present anywhere in the call stack.</span></span> <span data-ttu-id="8bfea-117">C# ' ta, `catch` anahtar sözcüğü bir özel durum işleyici tanımlamak için kullanılır.</span><span class="sxs-lookup"><span data-stu-id="8bfea-117">In C#, the `catch` keyword is used to define an exception handler.</span></span>  
  
-   <span data-ttu-id="8bfea-118">Belirli bir özel durum için hiçbir özel durum işleyicisi varsa, bir hata iletisi ile yürütme programı durdurur.</span><span class="sxs-lookup"><span data-stu-id="8bfea-118">If no exception handler for a given exception is present, the program stops executing with an error message.</span></span>  
  
-   <span data-ttu-id="8bfea-119">Bunu işlemek ve uygulama bilinen bir durumda bırakır sürece bir özel durum catch değil.</span><span class="sxs-lookup"><span data-stu-id="8bfea-119">Do not catch an exception unless you can handle it and leave the application in a known state.</span></span> <span data-ttu-id="8bfea-120">Catch `System.Exception`, kullanarak yeniden oluşturulması `throw` anahtar sözcüğü sonunda `catch` bloğu.</span><span class="sxs-lookup"><span data-stu-id="8bfea-120">If you catch `System.Exception`, rethrow it using the `throw` keyword at the end of the `catch` block.</span></span>  
  
-   <span data-ttu-id="8bfea-121">Varsa bir `catch` blok tanımlayan bir özel durum değişkeni, oluşan özel durum türü hakkında daha fazla bilgi edinmek için kullanın.</span><span class="sxs-lookup"><span data-stu-id="8bfea-121">If a `catch` block defines an exception variable, you can use it to obtain more information about the type of exception that occurred.</span></span>  
  
-   <span data-ttu-id="8bfea-122">Özel durumlar açıkça oluşturulabilir bir program tarafından kullanarak `throw` anahtar sözcüğü.</span><span class="sxs-lookup"><span data-stu-id="8bfea-122">Exceptions can be explicitly generated by a program by using the `throw` keyword.</span></span>  
  
-   <span data-ttu-id="8bfea-123">Özel durum nesneleri çağrı yığını ve metin açıklamasını hata durumu gibi hatayla ilgili ayrıntılı bilgiler içerir.</span><span class="sxs-lookup"><span data-stu-id="8bfea-123">Exception objects contain detailed information about the error, such as the state of the call stack and a text description of the error.</span></span>  
  
-   <span data-ttu-id="8bfea-124">Kod bir `finally` blok, bir özel durum olsa bile yürütüldüğünde.</span><span class="sxs-lookup"><span data-stu-id="8bfea-124">Code in a `finally` block is executed even if an exception is thrown.</span></span> <span data-ttu-id="8bfea-125">Kullanım bir `finally` herhangi bir akış veya açılmış dosyaları kapatmak örneğin kaynakları serbest bırakmak için blok `try` bloğu.</span><span class="sxs-lookup"><span data-stu-id="8bfea-125">Use a `finally` block to release resources, for example to close any streams or files that were opened in the `try` block.</span></span>  
  
-   <span data-ttu-id="8bfea-126">.NET Framework yönetilen özel durumlar mekanizması yapılandırılmış Win32 özel durum işleme üzerinde uygulanır.</span><span class="sxs-lookup"><span data-stu-id="8bfea-126">Managed exceptions in the .NET Framework are implemented on top of the Win32 structured exception handling mechanism.</span></span> <span data-ttu-id="8bfea-127">Daha fazla bilgi için bkz: [yapılandırılmış özel durum işleme (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) ve [bir kilitlenme indirmelere derinliklerine, Win32 yapılandırılmış özel durum işleme üzerinde](http://go.microsoft.com/fwlink/?LinkId=119654).</span><span class="sxs-lookup"><span data-stu-id="8bfea-127">For more information, see [Structured Exception Handling (C/C++)](/cpp/cpp/structured-exception-handling-c-cpp) and [A Crash Course on the Depths of Win32 Structured Exception Handling](http://go.microsoft.com/fwlink/?LinkId=119654).</span></span>  
  
## <a name="related-sections"></a><span data-ttu-id="8bfea-128">İlgili Bölümler</span><span class="sxs-lookup"><span data-stu-id="8bfea-128">Related Sections</span></span>  
 <span data-ttu-id="8bfea-129">Özel durumlar ve özel durum işleme hakkında daha fazla bilgi için aşağıdaki konulara bakın:</span><span class="sxs-lookup"><span data-stu-id="8bfea-129">See the following topics for more information about exceptions and exception handling:</span></span>  
  
-   [<span data-ttu-id="8bfea-130">Özel durumlar kullanma</span><span class="sxs-lookup"><span data-stu-id="8bfea-130">Using Exceptions</span></span>](../../../csharp/programming-guide/exceptions/using-exceptions.md)  
  
-   [<span data-ttu-id="8bfea-131">Özel durum işleme</span><span class="sxs-lookup"><span data-stu-id="8bfea-131">Exception Handling</span></span>](../../../csharp/programming-guide/exceptions/exception-handling.md)  
  
-   [<span data-ttu-id="8bfea-132">Oluşturma ve özel durumları atma</span><span class="sxs-lookup"><span data-stu-id="8bfea-132">Creating and Throwing Exceptions</span></span>](../../../csharp/programming-guide/exceptions/creating-and-throwing-exceptions.md)  
  
-   [<span data-ttu-id="8bfea-133">Derleyicinin ürettiği özel durumlar</span><span class="sxs-lookup"><span data-stu-id="8bfea-133">Compiler-Generated Exceptions</span></span>](../../../csharp/programming-guide/exceptions/compiler-generated-exceptions.md)  
  
-   [<span data-ttu-id="8bfea-134">Nasıl yapılır: bir özel durum try/catch kullanarak (C# programlama Kılavuzu) işleme</span><span class="sxs-lookup"><span data-stu-id="8bfea-134">How to: Handle an Exception Using try/catch (C# Programming Guide)</span></span>](../../../csharp/programming-guide/exceptions/how-to-handle-an-exception-using-try-catch.md)  
  
-   [<span data-ttu-id="8bfea-135">Nasıl yapılır: finally anahtar sözcüğünü kullanarak temizleme kodu yürütme</span><span class="sxs-lookup"><span data-stu-id="8bfea-135">How to: Execute Cleanup Code Using finally</span></span>](../../../csharp/programming-guide/exceptions/how-to-execute-cleanup-code-using-finally.md)  
  
## <a name="c-language-specification"></a><span data-ttu-id="8bfea-136">C# Dil Belirtimi</span><span class="sxs-lookup"><span data-stu-id="8bfea-136">C# Language Specification</span></span>  
 [!INCLUDE[CSharplangspec](~/includes/csharplangspec-md.md)]  
  
## <a name="see-also"></a><span data-ttu-id="8bfea-137">Ayrıca Bkz.</span><span class="sxs-lookup"><span data-stu-id="8bfea-137">See Also</span></span>  
 <xref:System.SystemException>  
 [<span data-ttu-id="8bfea-138">C# programlama kılavuzu</span><span class="sxs-lookup"><span data-stu-id="8bfea-138">C# Programming Guide</span></span>](../../../csharp/programming-guide/index.md)  
 [<span data-ttu-id="8bfea-139">C# anahtar sözcükleri</span><span class="sxs-lookup"><span data-stu-id="8bfea-139">C# Keywords</span></span>](../../../csharp/language-reference/keywords/index.md)  
 [<span data-ttu-id="8bfea-140">throw</span><span class="sxs-lookup"><span data-stu-id="8bfea-140">throw</span></span>](../../../csharp/language-reference/keywords/throw.md)  
 [<span data-ttu-id="8bfea-141">try-catch</span><span class="sxs-lookup"><span data-stu-id="8bfea-141">try-catch</span></span>](../../../csharp/language-reference/keywords/try-catch.md)  
 [<span data-ttu-id="8bfea-142">try-finally</span><span class="sxs-lookup"><span data-stu-id="8bfea-142">try-finally</span></span>](../../../csharp/language-reference/keywords/try-finally.md)  
 [<span data-ttu-id="8bfea-143">try-catch-finally</span><span class="sxs-lookup"><span data-stu-id="8bfea-143">try-catch-finally</span></span>](../../../csharp/language-reference/keywords/try-catch-finally.md)  
 [<span data-ttu-id="8bfea-144">Özel durumlar</span><span class="sxs-lookup"><span data-stu-id="8bfea-144">Exceptions</span></span>](../../../standard/exceptions/index.md)  
 [<span data-ttu-id="8bfea-145">Özel durum hiyerarşisi</span><span class="sxs-lookup"><span data-stu-id="8bfea-145">Exception Hierarchy</span></span>](http://msdn.microsoft.com/library/f7d68675-be06-40fb-a555-05f0c5a6f66b)  
 [<span data-ttu-id="8bfea-146">Güvenilir .NET kod yazma</span><span class="sxs-lookup"><span data-stu-id="8bfea-146">Writing Reliable .NET Code</span></span>](http://go.microsoft.com/fwlink/?LinkId=112400)  
 [<span data-ttu-id="8bfea-147">Mini dökümler belirli özel durumları</span><span class="sxs-lookup"><span data-stu-id="8bfea-147">Minidumps for Specific Exceptions</span></span>](http://go.microsoft.com/fwlink/?LinkId=112408)
